### PR TITLE
Fix wrong condition

### LINKED ISSUE #485: `gene_trends`only shows cells for the first lineage
I'm trying to plot gene expression trends along several lineages at the same time using 

```
model = cr.ul.models.GAMR(adata_raw, n_knots=5)

fig_kwargs = {'model': model,
              'data_key': 'magic_imputed_data', 
              'genes': ['Pdx1'], 
              'time_key': 'palantir_pseudotime', 
              'weight_threshold': (0.05, 0.05)
             }
cr.pl.gene_trends(adata_raw, **fig_kwargs)
```

This gives
image

For some reason, cells are only shown for the first lineage. I think that's a bug. 

#### Versions:
cellrank==1.2.0 scanpy==1.6.0 anndata==0.7.4 numpy==1.19.5 numba==0.52.0 scipy==1.5.3 pandas==1.1.3 scikit-learn==0.23.2 statsmodels==0.12.0 python-igraph==0.8.3 scvelo==0.2.2 pygam==0.8.0 matplotlib==3.2.2 seaborn==0.11.0
> ...

### LINKED ISSUE #486: `cell_color` parameter in `gene_trends` does not work as expected
In the API, is says: " Color of the cells when not visualizing absorption probabilities. Only used when same_plot=True". So I'm setting `same_plot=True` and I call:
```python
model = cr.ul.models.GAMR(adata_raw, n_knots=7)

fig_kwargs = {'model': model,
              'data_key': 'magic_imputed_data', 
              'genes': ['Pdx1'], 
              'same_plot': True,
             'cell_color': 'clusters',
              'time_key': 'palantir_pseudotime', 
              'weight_threshold': (0.05, 0.05)
             }

cr.pl.gene_trends(adata_raw, **fig_kwargs)
```
however, this does not show me the actual cells:
image

So I'm trying to get around this by explicitly passing `hide_cells=False`:
```python
model = cr.ul.models.GAMR(adata_raw, n_knots=7)

fig_kwargs = {'model': model,
              'data_key': 'magic_imputed_data', 
              'genes': ['Pdx1'], 
              'same_plot': True,
              'hide_cells': False,
              'cell_color': 'clusters',
              'time_key': 'palantir_pseudotime', 
              'weight_threshold': (0.05, 0.05)
             }

cr.pl.gene_trends(adata_raw, **fig_kwargs)
```
Doesn't make a difference: 
image

Let's remove the `cell_color` argument from the function call:
```python
model = cr.ul.models.GAMR(adata_raw, n_knots=7)

fig_kwargs = {'model': model,
              'data_key': 'magic_imputed_data', 
              'genes': ['Pdx1'], 
              'same_plot': True,
              'hide_cells': False,
              # 'cell_color': 'clusters',
              'time_key': 'palantir_pseudotime', 
              'weight_threshold': (0.05, 0.05)
             }

cr.pl.gene_trends(adata_raw, **fig_kwargs)
```
image

There are the cells, but now just shown in black. 

#### Versions:
cellrank==1.2.0 scanpy==1.6.0 anndata==0.7.4 numpy==1.19.5 numba==0.52.0 scipy==1.5.3 pandas==1.1.3 scikit-learn==0.23.2 statsmodels==0.12.0 python-igraph==0.8.3 scvelo==0.2.2 pygam==0.8.0 matplotlib==3.2.2 seaborn==0.11.0
> ...

### PR DESCRIPTION
**Pull request by michalk8**
**IMPORTANT: Please search among the [Pull request](../pulls) before creating one.**

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)

## How has this been tested?

TODO.

## Closes

closes #485 #486 #496

### CODE REVIEW

**Comment by michalk8** [comment]
@Marius1311 should be done as you've wanted, I've remove the old behavoir (found it useless).
Could you please test this out (i.e. try all corner cases: raw/layers, no colorbars, same_plot, missing, etc.) and maybe write tests?

**Comment by Marius1311** [comment]
> @Marius1311 should be done as you've wanted, I've remove the old behavoir (found it useless).
> Could you please test this out (i.e. try all corner cases: raw/layers, no colorbars, same_plot, missing, etc.) and maybe write tests?
> 
> Also, some olds tests should be modified/removed.

Hi @michalk8, I'll try this out next week! Thanks for fixing this.

**Comment by Marius1311** [comment]
Can confirm that this solves the color-problem I reported when plotting a single gene (via mattermost)

**Comment by Marius1311** [comment]
Can confirm that this fixes the `cell_color` problem I had and let's me color in `.obs` attributes.

**Comment by Marius1311** [comment]
I don't think `ncols` works as expected. See below, I set `ncols=1` but still get 4 columns:

```python
# using mgcv
model = cr.ul.models.GAMR(adata_raw, n_knots=n_knots_mgcv, smoothing_penalty=smoothing_penalty_mgcv)
    
fig_kwargs = {'model': model,
              'data_key': data_key, 
              'gene_as_title': False, 
              'legend_loc': None,
              'genes': ['Hhex'], 
              'ncols': 1,
              'time_key': pseudotime, 
              'same_plot': False, 
              'hide_cells': False,
              'weight_threshold': weight_threshold_mgcv, 
              'cell_color': 'clusters', 
              'figsize': (12, 3), }

cr.pl.gene_trends(adata_raw, **fig_kwargs)
```
gives
image

**Comment by Marius1311** [comment]
Would rename `cbar` to `colorbar` to stay close to scelo: https://scvelo.readthedocs.io/scvelo.pl.scatter.html#scvelo.pl.scatter

Would generally stay close to scVelo's scatter nomenclature.

**Comment by Marius1311** [comment]
Can confirm that turning the colorbar on/off works.

**Comment by Marius1311** [comment]
Can confirm that the original problem of #485 has been solved.

**Comment by Marius1311** [comment]
I don't think the `legend_loc` parameters works as expected - I basically just never get a legend when I color something in using `cell_color`:
```python
model = cr.ul.models.GAMR(adata_raw, n_knots=7)

fig_kwargs = {'model': model,
              'data_key': 'magic_imputed_data', 
              'genes': ['Pdx1'], 
              'time_key': 'palantir_pseudotime', 
              'weight_threshold': (0.05, 0.05),
              'cell_color': 'clusters',
              'legend_loc': 'right'
             }
cr.pl.gene_trends(adata_raw, **fig_kwargs)
```
gives
image

**Comment by Marius1311** [comment]
Sorry, this was my mistake: `legend_loc` **does** work, but only for lineages: 

```python
model = cr.ul.models.GAMR(adata_raw, n_knots=7)

fig_kwargs = {'model': model,
              'data_key': 'magic_imputed_data', 
              'genes': ['Pdx1'], 
              'time_key': 'palantir_pseudotime', 
              'weight_threshold': (0.05, 0.05),
              'cell_color': 'clusters',
              'legend_loc': 'best',
              'same_plot': True
             }
cr.pl.gene_trends(adata_raw, **fig_kwargs)
```
gives 
image
which is correct, of course.

**Comment by Marius1311** [comment]
We could need separate legends for lineages and cell-level `.obs` or `.varm` annotations passed via `cell_color`, similarly to how we do it when we plot a paga with pie charts.

**Comment by michalk8** [comment]
> I don't think ncols works as expected. See below, I set ncols=1 but still get 4 columns:

If you check the docstrings, `ncols` is only for `same_plot=True`, I never really intened it for `same_plot=False`, because I wanted a grid `genes x lineages`.

> We could need separate legends for lineages and cell-level .obs or .varm annotations passed via cell_color, similarly to how we do it when we plot a paga with pie charts.

Will be done soon as `obs_legend_loc`.

**Comment by michalk8** [comment]
@Marius1311 `obs_legend_loc` is implemented, also `cell_color` can now be passed specifically for `gene/lineage` combination as
```python
model = cr.ul.models.GAM(adata, n_knots=6)

fig_kwargs = {'model': model,
              'data_key': "Ms", 
              'gene_as_title': False, 
              'legend_loc': "best",
              'genes': ['Pdx1', adata.var_names[0]],
              'ncols': 1,
              'time_key': 'dpt_pseudotime', 
              'same_plot': False, 
              'hide_cells': False,
              'weight_threshold': (0.01, 0.01), 
              'cell_color': {"Pdx1": {"Alpha": "clusters", "Beta": "Pdx1", 'Epsilon': adata.var_names[3]},
                             "Snhg6": {"Alpha": "clusters_coarse"}},
              'transpose': True,
              'sharey': False,
              'suptitle': ''}

cr.pl.gene_trends(adata, **fig_kwargs, cbar=True)
```
yields:
index

note that this is really, really experimental (read: might be buggy - no tests for this).

**Comment by Marius1311** [comment]
Hi Mike, my feeling is that this is an overkill - if I really wanted to color cells differently per lineage/gene, I would just plot them separetely. The docstring is already super long and convoluted, I would try to keep things simple. I will check this out later and see whether it works on my data.

**Comment by Marius1311** [comment]
Comments:
- allowing `.var_names` annotations to be colored in via `cell_color` does not make sense: think about it, we're showing expression values of one gene, why would someone want to color the dots by another gene? I would vote for removing this function and only searching `.obs` for passed keywords
- setting `obs_legend_loc=None` does not get rid of the legend, this is unexpected. 
- I would **strongly** vote for only allowing one `cell_color` to be used across lineage and genes - this massively overcomplicates code & documentation. Currently, calling the code from example 1 gives nonsense because the legend should be shared across plots. Let's make this **simple** before we go into the details here.  

## Example 1
```
# using mgcv
model = cr.ul.models.GAMR(adata_raw, n_knots=n_knots_mgcv, smoothing_penalty=smoothing_penalty_mgcv)
    
fig_kwargs = {'model': model,
              'data_key': data_key, 
              'gene_as_title': False, 
              'legend_loc': None,
              'obs_legend_loc': 'best',
              'genes': ['Hhex'], 
              'ncols': 1,
              'time_key': pseudotime, 
              'same_plot': False, 
              'hide_cells': False,
              'weight_threshold': weight_threshold_mgcv, 
              'cbar': True, 
              'cell_color': 'clusters',
              'figsize': (12, 3), }

cr.pl.gene_trends(adata_raw, **fig_kwargs)
```
image

**Comment by michalk8** [comment]
> allowing .var_names annotations to be colored in via cell_color does not make sense: think about it, we're showing expression values of one gene, why would someone want to color the dots by another gene? I would vote for removing this function and only searching .obs for passed keywords

I disagree here, because we can have `data_key='obs'`.

> setting obs_legend_loc=None does not get rid of the legend, this is unexpected.

This will be fixed.

> I would strongly vote for only allowing one cell_color to be used across lineage and genes - this massively overcomplicates code & documentation. Currently, calling the code from example 1 gives nonsense because the legend should be shared across plots. Let's make this simple before we go into the details here. Similar to how we share the colorbar across subplots, the same should be done for the obs_legend.

Ok, then let's share it across plots and for categorical, show only legend in the last one.

> Why is there a colorbar shown in Example 1?

Bug, will be fixed.

> Need to enable obs_legend to be placed outside of plot, see Example 2. Currently, I can't even get rid of the legend that blocks everything. Please allow both None and "none", as done elsewhere. Also, same colorbar problem as in Example 1

This is already done, you can do it as described in `cellrank.pl._utils._position_legend`, e.g. `top right out`.

**Comment by Marius1311** [comment]
> > allowing .var_names annotations to be colored in via cell_color does not make sense: think about it, we're showing expression values of one gene, why would someone want to color the dots by another gene? I would vote for removing this function and only searching .obs for passed keywords
> 
> I disagree here, because we can have `data_key='obs'`.

You're right, I forgot about that. 

> 
> > setting obs_legend_loc=None does not get rid of the legend, this is unexpected.
> 
> This will be fixed.
> 
> > I would strongly vote for only allowing one cell_color to be used across lineage and genes - this massively overcomplicates code & documentation. Currently, calling the code from example 1 gives nonsense because the legend should be shared across plots. Let's make this simple before we go into the details here. Similar to how we share the colorbar across subplots, the same should be done for the obs_legend.
> 
> Ok, then let's share it across plots and for categorical, show only legend in the last one.
> 
> > Why is there a colorbar shown in Example 1?
> 
> Bug, will be fixed.
> 
> > Need to enable obs_legend to be placed outside of plot, see Example 2. Currently, I can't even get rid of the legend that blocks everything. Please allow both None and "none", as done elsewhere. Also, same colorbar problem as in Example 1
> 
> This is already done, you can do it as described in `cellrank.pl._utils._position_legend`, e.g. `top right out`.

Great, thanks! Let me know when I should test this again.

**Comment by michalk8** [comment]
You can test this again, the legend should be visible only once, in the last plot of the 1st row.

**Comment by Marius1311** [comment]
Great, will take a look!

**Comment by Marius1311** [comment]
> You can test this again, the legend should be visible only once, in the last plot of the 1st row.

yes! I can confirm that this works.

**Comment by Marius1311** [comment]
This PR solves the original two issues, which were
- allowing passing `.obs` keywords to the `cell_color` parameter (#486)
- showing cells in all lineages if `same_plot=False` and `hide_cells=False` (#485)

There are still a couple of things with legend placement and colorbars I'm looking into at the moment which could be improved, but I think that should be a separate issue/PR to keeps things tidy. So I would vote for fixing the tests, merging this, and then moving on to these other issues.

**Review by Marius1311 — APPROVED** [review]
Very nice work, once the tests are fixed, this can be merged.

File: cellrank/pl/_gene_trend.py [hunk 1]
```diff
@@ -134,6 +134,8 @@ def gene_trends(
         Whether to show gene names as titles instead on y-axis.
     legend_loc
         Location of the legend displaying lineages. Only used when `same_plot=True`.
+    obs_legend_loc
```
[the remark this answers is not in the document]
Comment (michalk8) [reply cellrank/pl/_gene_trend.py:137]: I agree, let's use `lineage_legend_loc` and `obs_legend_loc`.
The only alternative would be to merge `cbar` and `obs_legend_loc` into 1 argument, since they are mutually exclusive, though I am not sure what the name would be.

[the remark this answers is not in the document]
Comment (Marius1311) [reply cellrank/pl/_gene_trend.py:137]: But `cbar` is `bool`, no? Let's leave them separate I would say.

**Comment by michalk8** [comment]
Ok, all tests pass, you can now also put the lineage legend outside.

**Comment by Marius1311** [comment]
Very nice work, thanks a lot!
